### ISSUE TITLE
Timing of MainLoop.Invoke()

### ISSUE THREAD

**Issue by brooke-hamilton** [migueldeicaza/gui.cs#155]
I have a background task that is updating the UI. I realize this is not the preferred approach ([Issue 145](https://github.com/migueldeicaza/gui.cs/issues/145)), and maybe someone can suggest a better one. However, I'm following the guidelines on this page [https://migueldeicaza.github.io/gui.cs/articles/mainloop.html#threading
](https://migueldeicaza.github.io/gui.cs/articles/mainloop.html#threading
) by using the `Application.MainLoop.Invoke()` method to run code that updates the UI.

`MainLoop.Invoke()` works exactly as described, except for one thing I did not expect: the timing of the action being invoked seems to depend on the next time the main loop wakes up to do some processing. For example, moving the mouse around results in the actions passed to `MainLoop.Invoke()` being executed, but without some user manipulation to wake up the loop, the actions wait in the queue.  Is that expected or is it intentional?  

I came up with a workaround, which is to wake up the main loop every 500 milliseconds with a timeout that does nothing but return true.  Is there a better way?  See repro below.

I'm running this on Windows 10 version 1809.

```CS
private static void Main(string[] args)
{
    Application.Init();

    var messages = new List<string>();
    var listView = new ListView(messages);
    Application.Top.Add(listView);

    // This action is runs continuously in the background. It represents a an application that is
    // subscribing to a message queue and displaying the messages on the UI. This is done on a separate
    // thread because the connection with the remote queue is persistent, and the client library method
    // call to retrieve the next message blocks until the next message arrives.
    Task.Run(() =>
    {
        int x = 1;
        while (true) // This would normally be waiting for a cancellation token to request cancellation.
        {
            messages.Add($"Message {x++}");
            while (messages.Count > 25) messages.RemoveAt(0); // Limit the list to 25 messages.

            // Require repainting of child views in the list view.
            Application.MainLoop.Invoke(() => listView.ChildNeedsDisplay());
            // Should MainLoop wake up right now to process the action above? Or another method call?

            Thread.Sleep(1000); // Pretend this is a blocking method call waiting for the next message.
        }
    });

    // This timeout does nothing but return true, however it appears to trigger the Application.MainLoop
    // to run the Action above that calls listView.ChildNeedsDisplay(). Without this timeout, the display
    // updates are random, or triggered by user interaction with the UI. Seems like a hack.
    Application.MainLoop.AddTimeout(TimeSpan.FromMilliseconds(500), loop => true);

    Application.Run();
}

```

Love this framework, by the way.

**Comment by migueldeicaza** [comment]
Good catch, this has to be a bug on the Windows driver, as any "Invoke" really should trigger a wakeup from the mainloop.

**Comment by INeedAUniqueUsername** [comment]
I'm getting the same problem from MainLoop.AddIdle for Windows.

**Comment by brooke-hamilton** [comment]
I believe this issue is resolved with PR #208.
